firebase-ai/src/main/kotlin/com/google/firebase/ai/TemplateChat.kt

```diff
@@ -0,0 +1,166 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.ai
+
+import android.graphics.Bitmap
+import com.google.firebase.ai.type.Content
+import com.google.firebase.ai.type.GenerateContentResponse
+import com.google.firebase.ai.type.ImagePart
+import com.google.firebase.ai.type.InlineDataPart
+import com.google.firebase.ai.type.InvalidStateException
+import com.google.firebase.ai.type.PublicPreviewAPI
+import com.google.firebase.ai.type.TextPart
+import com.google.firebase.ai.type.content
+import java.util.LinkedList
+import java.util.concurrent.Semaphore
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.onEach
+
+/**
+ * Representation of a multi-turn interaction with a model.
+ *
+ * Captures and stores the history of communication in memory, and provides it as context with each
+ * new message.
+ *
+ * **Note:** This object is not thread-safe, and calling [sendMessage] multiple times without
+ * waiting for a response will throw an [InvalidStateException].
+ *
+ * @param model The model to use for the interaction.
+ * @param templateID The template ID for this chat session
+ * @property history The previous content from the chat that has been successfully sent and received
+ * from the model. This will be provided to the model for each message sent (as context for the
+ * discussion).
+ */
+@PublicPreviewAPI
+public class TemplateChat(
+  private val model: TemplateGenerativeModel,
+  private val templateID: String,
+  public val history: MutableList<Content> = ArrayList()
+) {
+  private var lock = Semaphore(1)
+
+  /**
+   * Sends a message using the provided [prompt]; automatically providing the existing [history] as
+   * context.
+   *
+   * If successful, the message and response will be added to the [history]. If unsuccessful,
+   * [history] will remain unchanged.
+   *
+   * @param prompt The input that, together with the history, will be given to the model as the
+   * prompt.
+   * @param inputs the inputs needed to fill in the template ID
+   * @throws InvalidStateException if [prompt] is not coming from the 'user' role.
+   * @throws InvalidStateException if the [Chat] instance has an active request.
+   */
+  public suspend fun sendMessage(
+    prompt: Content,
+    inputs: Map<String, Any>
+  ): GenerateContentResponse {
+    prompt.assertComesFromUser()
+    attemptLock()
+    try {
+      val fullPrompt = history + prompt
+      val response = model.generateContent(templateID, inputs, fullPrompt)
```

Comment:
This seems different from what we discussed offline. In here the prompt is being send to the server